IPython/html/widgets/widget_output.py

```diff
@@ -38,13 +39,41 @@ def clear_output(self, *pargs, **kwargs):
             clear_output(*pargs, **kwargs)
 
     def __enter__(self):
+        """Called upon entering output widget context manager."""
         self._flush()
-        self.send({'method': 'push'})
+        kernel = get_ipython().kernel
+        session = kernel.session
+        send = session.send
+        self._original_send = send
+        self._session = session
+
+        def send_hook(stream, msg_or_type, content=None, parent=None, ident=None,
+             buffers=None, track=False, header=None, metadata=None): 
+
+            # Handle both prebuild messages and unbuilt messages.
+            if isinstance(msg_or_type, (Message, dict)):
+                msg_type = msg_or_type['msg_type']
+                msg = dict(msg_or_type)
+            else:
+                msg_type = msg_or_type
+                msg = session.msg(msg_type, content=content, parent=parent, 
+                    header=header, metadata=metadata)
+
+            # If this is a message type that we want to forward, forward it.
+            if stream is kernel.iopub_socket and msg_type in ['clear_output', 'stream', 'display_data']:
+                self.send(msg)
+            else: 
+                send(stream, msg_or_type, content=content, parent=parent, ident=ident,
+                    buffers=buffers, track=track, header=header, metadata=metadata)
```

Comment:
Is it more efficient to use the `msg` that is already prepared, rather than potentially making session build it again? i.e. call this as `send(stream, msg, ident=ident, buffers=buffers, track=track)`?